File: base62.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

static char * base62_digits="0123456789"
                            "abcdefghijklmnopqrstuvwxyz"
                            "ABCDEFGHIJKLMNOPQRSTUVWXYZ" ;

char * b62_encode(uint64_t num)
{
    /* 64 bits is up to 11 digits in base 62 */
    static char enc[15] ;
    int         i, len ;
    uint64_t    rem ;
    char        c ;

    /* Zero out output buffer */
    memset(enc, 0, sizeof(enc));
    if (num==0) {
        enc[0]='0';
        return enc;
    }
    i=0 ;
    while (num>0) {
        rem = num % 62 ;
        enc[i] = base62_digits[rem];
        i++ ;
        num /= 62 ;
    }
    /* Reverse string */
    len = (int)strlen(enc);
    for (i=0 ; i<len/2 ; i++) {
        c = enc[i] ;
        enc[i] = enc[len-i-1];
        enc[len-i-1] = c ;
    }
    return enc ;
}

uint64_t b62_decode(char * s)
{
    uint64_t    num ;
    uint64_t    p ;
    char     *  pos ;
    int         i ;

    if (!s)
        return 0 ;

    p=1;
    num=0;
    for (i=strlen(s)-1 ; i>=0 ; i--) {
        pos = strchr(base62_digits, (int)s[i]);
        if (!pos) {
            return 0 ;
        }
        num += (uint64_t)(pos-base62_digits) * p ;
        p *= 62 ;
    }
    return num ;
}

#ifdef MAIN
int main(int argc, char * argv[])
{
    uint64_t num ;
    char * s ;

	if (argc<2) {
		printf("use: %s\n", argv[0]);
		return 1 ;
	}

    sscanf(argv[1], "%llu", &num);
    s   = b62_encode(num);
    printf("base62: %s\n", s);
    num = b62_decode(s);
    printf("dec: %llu\n", num);
	return 0 ;
}
#endif

File: base62.go
package main

import (
    "fmt"
    "os"
    "strconv"
    "strings"
)

const (
    b62_digits =
    "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
)

func b62_encode(num uint64) string {
    var rem uint64
    s:=""
    for num>0 {
        rem = num % 62 ;
        s=string(b62_digits[rem])+s
        num = num / 62
    }
    return s
}

func b62_decode(s string) uint64 {
    var (
        num uint64
        pow uint64
    )
    pow=1
    for i:=len(s)-1 ; i>=0 ; i-- {
        num += pow * uint64(strings.Index(b62_digits, string(s[i])))
        pow *= 62
    }
    return num
}

func main() {
    if len(os.Args)<2 {
        fmt.Printf("use: %s NUM\n", os.Args[0])
        return
    }
    val, _ := strconv.ParseUint(os.Args[1], 10, 64)
    enc := b62_encode(val)
    fmt.Printf("b62[%s]\n", enc)
    dec := b62_decode(enc)
    fmt.Printf("-> %v\n", dec)
}

